tracker/src/index.rs

```diff
@@ -0,0 +1,331 @@
+use crate::bitcoin_client::BitcoinClient;
+use crate::record::{Record, RecordData};
+use crate::storage::IndexStorage;
+use bitcoin::blockdata::opcodes;
+use bitcoin::blockdata::script::Instruction;
+use bitcoin::{BlockHash, Transaction, TxOut, Txid};
+use bitcoincore_rpc::json::GetBlockHeaderResult;
+use std::cmp::Ordering;
+use std::convert::TryInto;
+
+pub type BagId = [u8; 32];
+
+pub struct Index<C: BitcoinClient, S: IndexStorage> {
+    client: C,
+    height: u64,
+    checked_chain: Vec<BlockHash>,
+    storage: S,
+    bags: Vec<BagId>,
+}
+
+impl<C: BitcoinClient, S: IndexStorage> Index<C, S> {
+    pub fn new(client: C, storage: S, base_height: Option<u64>) -> Self {
+        let info = client.get_blockchain_info().unwrap();
+        let height = base_height.unwrap_or(info.blocks);
+        let checked_chain = Vec::new();
+        let bags = vec![];
+        Index {
+            client,
+            height,
+            checked_chain,
+            storage,
+            bags,
+        }
+    }
+
+    pub fn checked_height(&self) -> u64 {
+        self.height
+    }
+
+    pub fn add_bag(&mut self, bag: BagId) {
+        self.bags.push(bag);
+    }
+
+    // TODO: better API for indexing
+    pub fn get_storage(&self) -> &S {
+        &self.storage
+    }
+}
+
+impl<C: BitcoinClient, S: IndexStorage> Index<C, S> {
+    pub fn check_last_blocks(&mut self) {
+        let new_blockchain_info = self.client.get_blockchain_info().unwrap();
+        let new_height = new_blockchain_info.blocks;
+        match self.height.cmp(&new_height) {
+            Ordering::Equal => return,
+            Ordering::Greater => {
+                unimplemented!("TODO: is it possible that chain will be less than previous?")
```

Comment:
It could be in case of a deep or across-retarget reorg. Difficulty and each block’s weight is fixed within each retarget interval (2016 blocks, ≈2 wks). So within an interval, with common ancestor in the same interval, a chain with lower height is strictly lower in weight. Otherwise, not.